File: app/src/main/java/com/capstone/bankit/ui/adapters/AnalyticsAdapter.java
package com.capstone.bankit.ui.adapters;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.capstone.bankit.R;
import com.capstone.bankit.data.models.TransactionModel;
import com.capstone.bankit.databinding.ItemAnalyticsRowBinding;
import com.capstone.bankit.databinding.ItemTransactionsRowBinding;
import com.capstone.bankit.utils.Constants;

import java.util.Objects;


public class AnalyticsAdapter extends ListAdapter<TransactionModel, AnalyticsAdapter.ViewHolder> {
    public AnalyticsAdapter() {
        super(DIFF_CALLBACK);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemAnalyticsRowBinding binding = ItemAnalyticsRowBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false
        );
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        TransactionModel transactionModel = getItem(position);
        if (transactionModel != null) {
            String amount = Constants.formatToRupiah(transactionModel.getAmount());
            int image = Constants.mapTransactionTypeToIcon(transactionModel.getType());

            holder.binding.ivTransaction.setImageDrawable(ContextCompat.getDrawable(holder.binding.getRoot().getContext(), image));
            holder.binding.tvType.setText(transactionModel.getType());
            holder.binding.tvPercentage.setText(transactionModel.getPercentage() + "%");
            holder.binding.tvAmount.setText(amount);
        }
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        final ItemAnalyticsRowBinding binding;

        public ViewHolder(@NonNull ItemAnalyticsRowBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    private static final DiffUtil.ItemCallback<TransactionModel> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<TransactionModel>() {
                @Override
                public boolean areItemsTheSame(@NonNull TransactionModel oldStory,
                                               @NonNull TransactionModel newStory) {
                    return oldStory.equals(newStory);
                }

                @Override
                public boolean areContentsTheSame(@NonNull TransactionModel oldStory,
                                                  @NonNull TransactionModel newStory) {
                    return Objects.equals(oldStory.getId(), newStory.getId());
                }
            };
}

File: app/src/main/java/com/capstone/bankit/ui/adapter/AnalyticsAdapter.kt
package com.capstone.bankit.ui.adapter

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.capstone.bankit.databinding.ItemAnalyticsRowBinding
import com.capstone.bankit.utils.Constants.formatToRupiah
import com.capstone.bankit.utils.Constants.mapTransactionTypeToIcon
import com.capstone.bankit.data.models.DataItem

class AnalyticsAdapter : ListAdapter<DataItem, AnalyticsAdapter.ViewHolder>(DIFF_CALLBACK) {

    private var totalAmount: Double = 0.0

    @SuppressLint("NotifyDataSetChanged")
    fun setTotalAmount(amount: Double) {
        totalAmount = amount
        notifyDataSetChanged() // Ensure that the percentages are updated
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemAnalyticsRowBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val dataItem = getItem(position)

        // Make sure dataItem is not null
        if (dataItem != null) {
            // Safely extract category and amount
            val category = dataItem.category ?: "Unknown"
            val amount = dataItem.amount?.let { formatToRupiah(it) } ?: "0"
            val image = mapTransactionTypeToIcon(category)

            // Set icon for transaction type (if category is known)
            holder.binding.ivTransaction.setImageDrawable(
                ContextCompat.getDrawable(holder.binding.root.context, image)
            )

            // Calculate percentage if totalAmount is greater than 0
            val percentage = if (totalAmount > 0) {
                val categoryAmount = dataItem.amount ?: 0.0
                ((categoryAmount / totalAmount) * 100).toInt()
            } else {
                0
            }

            // Update UI
            holder.binding.tvType.text = category
            holder.binding.tvPercentage.text = "$percentage%" // Display percentage
            holder.binding.tvAmount.text = amount
        }
    }

    class ViewHolder(val binding: ItemAnalyticsRowBinding) : RecyclerView.ViewHolder(binding.root)

    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<DataItem> =
            object : DiffUtil.ItemCallback<DataItem>() {
                override fun areItemsTheSame(oldItem: DataItem, newItem: DataItem): Boolean {
                    return oldItem.id == newItem.id
                }

                override fun areContentsTheSame(oldItem: DataItem, newItem: DataItem): Boolean {
                    return oldItem == newItem
                }
            }
    }
}
